modules/calib3d/src/solvepnp.cpp

```diff
@@ -456,4 +398,325 @@ int solveP3P( InputArray _opoints, InputArray _ipoints,
+            CV_DbgCheckEQ(Xs[3][0], Xs[3][1], "Object points must be: X3 == Y3!");
+            return false;
+        }
+
+        Mat undistortedPoints;
+        undistortPoints(ipoints, undistortedPoints, cameraMatrix, distCoeffs);
+
+        IPPE::PoseSolver poseSolver;
+        Mat rvec1, tvec1, rvec2, tvec2;
+        float reprojErr1, reprojErr2;
+        poseSolver.solveSquare(opoints, undistortedPoints, rvec1, tvec1, reprojErr1, rvec2, tvec2, reprojErr2);
+
+        if (reprojErr1 < reprojErr2)
+        {
+            vec_rvecs.push_back(rvec1);
+            vec_tvecs.push_back(tvec1);
+
+            vec_rvecs.push_back(rvec2);
+            vec_tvecs.push_back(tvec2);
+        }
+        else
+        {
+            vec_rvecs.push_back(rvec2);
+            vec_tvecs.push_back(tvec2);
+
+            vec_rvecs.push_back(rvec1);
+            vec_tvecs.push_back(tvec1);
+        }
+    }
+    /*else if (flags == SOLVEPNP_DLS)
+    {
+        Mat undistortedPoints;
+        undistortPoints(ipoints, undistortedPoints, cameraMatrix, distCoeffs);
+
+        dls PnP(opoints, undistortedPoints);
+
+        Mat rvec, tvec, R;
+        bool result = PnP.compute_pose(R, tvec);
+        if (result)
+        {
+            Rodrigues(R, rvec);
+            vec_rvecs.push_back(rvec);
+            vec_tvecs.push_back(tvec);
+        }
+    }
+    else if (flags == SOLVEPNP_UPNP)
+    {
+        upnp PnP(cameraMatrix, opoints, ipoints);
+
+        Mat rvec, tvec, R;
+        PnP.compute_pose(R, tvec);
+        Rodrigues(R, rvec);
+        vec_rvecs.push_back(rvec);
+        vec_tvecs.push_back(tvec);
+    }*/
+    else
+        CV_Error(CV_StsBadArg, "The flags argument must be one of SOLVEPNP_ITERATIVE, SOLVEPNP_P3P, SOLVEPNP_EPNP or SOLVEPNP_DLS");
+
+    CV_Assert(vec_rvecs.size() == vec_tvecs.size());
+
+    int solutions = static_cast<int>(vec_rvecs.size());
+    _rvecs.create(solutions, 1, mtype);
+    _tvecs.create(solutions, 1, mtype);
+
+    for (int i = 0; i < solutions; i++)
+    {
+        _tvecs.getMatRef(i) = vec_tvecs[i];
+        _rvecs.getMatRef(i) = vec_rvecs[i];
+    }
+
+    if (reprojectionError.needed())
+    {
+        int type = reprojectionError.type();
+        reprojectionError.create(solutions, 1, type);
+        CV_CheckType(reprojectionError.type(), type == CV_32FC1 || type == CV_64FC1,
+                     "Type of reprojectionError must be CV_32FC1 or CV_64FC1!");
+
+        Mat objectPoints, imagePoints;
+        if (_opoints.depth() == CV_32F)
+        {
+            _opoints.getMat().convertTo(objectPoints, CV_64F);
+        }
+        else
+        {
+            objectPoints = _opoints.getMat();
+        }
+        if (_ipoints.depth() == CV_32F)
+        {
+            _ipoints.getMat().convertTo(imagePoints, CV_64F);
+        }
+        else
+        {
+            imagePoints = _ipoints.getMat();
+        }
+
+        for (size_t i = 0; i < vec_rvecs.size(); i++)
```

Comment:
Here the reprojection errors are computed. Some methods already computed the reprojection error but for consistency (e.g. in px?, in m in the normalized camera frame?, ...) I think it is better to do it at the end.

Reply:
```
if (err.needed())
{
    ... fill err ...
}
```

Reply:
The check should be there, just above.

Reply:
Right!
Lets leave `err` update as is.